### ISSUE TITLE
titan-server usability fixes

### ISSUE THREAD

**Issue by dalaro** [thinkaurelius/titan#463]
Collecting recurring issues from the list and SO:
- Make titan.sh wait for Cassandra to fully bootstrap so that Astyanax doesn't try to talk to it before it's completely initialized; I want to avoid introducing a command dependency, so that probably implies either log watching or using one of Cassandra's stock tools (probably nodetool)
- Make titan.sh process its verb only after all options have been completely processed.  In 0.4.0, options must be specified before the verb to take effect, which is unintuitive and violation of the least-surprise principle.
- Check CLASSPATH separator (: vs ;) on Windows for Cygwin
- Add a switch to control Rexster/Cassandra log spew to console
- Make Rexster actually log to log/rexstitan.log instead of just to the console; setting -Dlog4j.configuration=conf/log4j-rexstitan.properties does not seem to have an effect -- maybe Rexster overrides the default Log4j init logic?

Things that are already fixed in master:
- Add Rexster's public web UI assets to the root of the distro and make sure Rexster can find them
- Make titan.sh generate paths/classpaths correctly regardless of the user's current working directory

**Comment by mbroecheler** [comment]
@spmallette Can you comment on the log file configuration?

**Comment by spmallette** [comment]
Rexster's log file configuration is more complicated than I would like, but here goes.  By default rexster suppresses rediculous amounts of log output from jersey and grizzly.  To flip that on you pass a `-d` on the command line.  To have better control over the logging output itself, Rexster looks for a `log4j.properties` file.  By default it uses the one  on the classpath embedded in one of the rexster jars, but will also look for one in the root of the rexster working directory as an override.  I'm not sure but I think all of that can be overridden by passing something like `-Dlog4j.configuration=file:config/log4j.properties` that specifies the location of the file.

**Comment by dalaro** [comment]
These items are done in master:
- Make titan.sh wait for Cassandra to fully bootstrap so that Astyanax doesn't try to talk to it before it's completely initialized; I want to avoid introducing a command dependency, so that probably implies either log watching or using one of Cassandra's stock tools (probably nodetool) (58bc8358459f51acb836fc0b16e5aea01cd3fe9a)
- Make titan.sh process its verb only after all options have been completely processed.  In 0.4.0, options must be specified before the verb to take effect, which is unintuitive and violation of the least-surprise principle. (3d5cd857034b319ea3239ac4c6e854ac42163864)
- Add a switch to control Rexster/Cassandra log spew to console (3d5cd857034b319ea3239ac4c6e854ac42163864 disables console spew by default, use "-v" to enable)
- Make Rexster actually log to log/rexstitan.log instead of just to the console; setting -Dlog4j.configuration=conf/log4j-rexstitan.properties does not seem to have an effect -- maybe Rexster overrides the default Log4j init logic? (375d71b7d5f7143f5838ae2d8d70f91d75298606)

This leaves one item, currently in progress as I test and patch titan-server on Windows 7 + Cygwin:
- Check CLASSPATH separator (: vs ;) on Windows for Cygwin

**Comment by dalaro** [comment]
Re Rexster logging.

I started out with -Dlog4j.configuration=conf/log4j-rexstitan.properties.  This props file configured logging to a file.  The file would be created, but it would remain empty.  I looked at Rexster's Application class.  Because the static initializer always runs PropertyConfigurator.configure(...) on either a file or classpath resource named "log4j.properties", and because I don't see a way to override that name, I think what happens is that log4j's builtin static initializers read -Dlog4j.configuration=conf/log4j-rexstitan.properties, created the file, and then Application#main() ran pretty much right afterward and overwrote that configuration with some log4j.properties from the classpath.  I ended up making a dedicated working directory for Rexster containing its log4j.properties file (formerly conf/log4j-rexstitan.properties), and that appears to work.

**Comment by spmallette** [comment]
That sounds right.  I probably should change that to better respect `-Dlog4j.configuration` at some point, but since you figured a work around i guess that will work.

**Comment by dalaro** [comment]
f8404ceefea2be717bc24ccc6d6dd3ef865acca2 adds Cygwin compatibility to the titan-server distro.  That patch basically does two things based on the value of `uname`:
- Invoke `cygpath` to convert *NIX paths to Windows equivalents when needed
- Call `taskkill` instead of `kill` on Cygwin

titan.sh start, stop, status, and clean worked as expected on my Win 7 test box under current Cygwin with these changes.  I briefly tested the same on Linux and found no regressions.

However, I have an unsolved problem involving rexster-console.sh on Cygwin.  It prints the usual welcome banner, logs the creation of a context, and presents the command prompt.  I type a command and the console echoes it verbatim.  So far so good.  But when I hit enter, rexster-console appears to hang without evaluating the command.  Even if the expression is just "1 + 1", it does not seem to return.  This happened with both the rexster-console.sh file in the Titan tree and the one in rexster-console-2.4.0.

For the current release, I've added rexster-console.bat to the titan-server distro.  This script works as expected when run through the Windows command prompt.  Because rexster-console.bat works on Cygwin where rexster-console.sh does not, I suspect this is some sort of mismatch between Cygwin's terminal implementation and Rexster's I/O handling, but I'm only guessing right now.

**Comment by spmallette** [comment]
I think it would be worth removing some unnecessary files from the distribution (specifically from the `conf` directory) so as to avoid confusion as to what files are actually used.  I'm not sure what's "necessary" or isn't but I would guess that `titan-server-rexster.xml` isn't used as well as non-cassandra/-es properties files.  I've caught myself editing the wrong "rexster.xml" in there and it's fooled a few people on the mailing list.

**Comment by dalaro** [comment]
That's a good point.  I'm going to move config file restructuring into a separate issue, since this one is starting to get into mission creep.

**Comment by dalaro** [comment]
#587
